### PR TITLE
Verify send guard in Run

### PR DESCRIPTION
**Pull request by mlimbeck**
This PR supersedes PR #249.

It is currently blocked, as it depends on https://github.com/viperproject/gobra/pull/749.

It contains the current progress of verifying the send guard in the run method.
Still missing:
- [x] Proof of `absIO_valWidenLemma` (#268)
- [x] Fix verification errors in Run
- [x] use `--moreJoins`

### CODE REVIEW

**Comment by jcp19** [comment]
I did a bit more debugging here. I believe that there is an incompleteness in Silicon that I reported to @marcoeilers

File: router/dataplane.go [hunk 1]
```diff
@@ -1358,13 +1430,15 @@ func (p *scionPacketProcessor) reset() (err error) {
...
-// @	newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)
-// TODO: On a first step, we will prove that whenever we have a valid scion packet in processSCION,
-// the correct "next packet" is computed
+// @ 	absPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.IBufY, ifsToIO_ifs(p.getIngressID()), absPkt.IO_val_Pkt2_2)
+// @ ensures dp.Valid()
+// @ ensures reserr == nil ==>
+// @ 	respr.OutPkt != nil                                      &&
+// @ 	newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID) &&
+// @ 	newAbsPkt.isIO_val_Pkt2                                  &&
+// @ 	ElemWitness(ioSharedArg.OBufY, newAbsPkt.IO_val_Pkt2_1, newAbsPkt.IO_val_Pkt2_2)
```
Comment (mlimbeck) [inline router/dataplane.go:1439]: Is this the implication `reserr == nil ==> newAbsPkt.isIO_val_Pkt2 ` correct? 
We could return an `IO_val_Unsupported` even if `reserr == nil`, e.g. in the epic case.

Comment (jcp19) [reply router/dataplane.go:1439]: oops, that is a good point!

File: router/dataplane.go [hunk 2]
```diff
@@ -1650,8 +1724,10 @@ func (p *scionPacketProcessor) processIntraBFD(data []byte) (res error) {
 // @	absPkt.isIO_val_Pkt2 ==> ElemWitness(ioSharedArg.IBufY, ifsToIO_ifs(p.ingressID), absPkt.IO_val_Pkt2_2)
 // @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==>
 // @	ElemWitness(ioSharedArg.OBufY, newAbsPkt.IO_val_Pkt2_1, newAbsPkt.IO_val_Pkt2_2)
-// @ ensures reserr == nil && newAbsPkt.isIO_val_Pkt2 ==> respr.OutPkt != nil &&
+// @ ensures respr.OutPkt != nil ==>
 // @	newAbsPkt == absIO_val(dp, respr.OutPkt, respr.EgressID)
+// @ ensures reserr != nil && respr.OutPkt != nil ==>
+// @ 	newAbsPkt.isIO_val_Unsupported
 func (p *scionPacketProcessor) processSCION( /*@ ghost ub []byte, ghost llIsNil bool, ghost startLL int, ghost endLL int, ghost ioLock *sync.Mutex, ghost ioSharedArg SharedArg, ghost dp io.DataPlaneSpec @*/ ) (respr processResult, reserr error /*@ , addrAliasesPkt bool, ghost newAbsPkt io.IO_val  @*/) {
```
Comment (mlimbeck) [inline router/dataplane.go:1731]: I think we need `ensures (respr.OutPkt == nil) == (newAbsPkt == io.IO_val_Unit{}) ` here as well?

**Pull request merged**